// dmn-js-decision-table/lib/features/decision-table-head/editor/components/OutputCell.js
function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }

import { createVNode } from "inferno";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); Object.defineProperty(subClass, "prototype", { writable: false }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } else if (call !== void 0) { throw new TypeError("Derived constructors may only return object or undefined"); } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import { Component } from 'inferno';
import { mixin } from 'table-js/lib/components';
import { ComponentWithSlots } from 'dmn-js-shared/lib/components/mixins';

var OutputCell = /*#__PURE__*/function (_Component) {
  _inherits(OutputCell, _Component);

  var _super = _createSuper(OutputCell);

  function OutputCell(props, context) {
    var _this;

    _classCallCheck(this, OutputCell);

    _this = _super.call(this, props, context);

    _defineProperty(_assertThisInitialized(_this), "onClick", function (event) {
      var output = _this.props.output;

      _this._eventBus.fire('output.edit', {
        event: event,
        output: output
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onContextmenu", function (event) {
      var id = _this.props.output.id;

      _this._eventBus.fire('cell.contextmenu', {
        event: event,
        id: id
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onElementsChanged", function () {
      _this.forceUpdate();
    });

    mixin(_assertThisInitialized(_this), ComponentWithSlots);
    _this._translate = context.injector.get('translate');
    return _this;
  }

  _createClass(OutputCell, [{
    key: "componentWillMount",
    value: function componentWillMount() {
      var injector = this.context.injector;
      this._changeSupport = this.context.changeSupport;
      this._eventBus = injector.get('eventBus');
      this._elementRegistry = injector.get('elementRegistry');
      var output = this.props.output;

      this._changeSupport.onElementsChanged(output.id, this.onElementsChanged);
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      var output = this.props.output;

      this._changeSupport.offElementsChanged(output.id, this.onElementsChanged);
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          output = _this$props.output,
          index = _this$props.index,
          outputsLength = _this$props.outputsLength;
      var label = output.label,
          name = output.name,
          outputValues = output.outputValues,
          typeRef = output.typeRef;
      var width = output.width ? output.width + 'px' : '192px';
      return createVNode(1, "th", "output-cell output-editor", [this.slotFills({
        type: 'cell-inner',
        context: {
          cellType: 'output-cell',
          col: this._elementRegistry.get(output.id),
          index: index,
          outputsLength: outputsLength
        },
        col: output
      }), createVNode(1, "div", "clause", index === 0 ? this._translate('Then') : this._translate('And'), 0), label ? createVNode(1, "div", "output-label", label, 0, {
        "title": this._translate('Output Label')
      }) : createVNode(1, "div", "output-name", name, 0, {
        "title": this._translate('Output Name')
      }), createVNode(1, "div", "output-variable", outputValues && outputValues.text || typeRef, 0, {
        "title": outputValues && outputValues.text ? this._translate('Output Values') : this._translate('Output Type')
      })], 0, {
        "data-col-id": output.id,
        "onDblClick": this.onClick,
        "onContextmenu": this.onContextmenu,
        "style": {
          width: width
        }
      });
    }
  }]);

  return OutputCell;
}(Component);

export { OutputCell as default };
